samigo/samigo-impl/src/java/org/sakaiproject/samigo/impl/SamigoAvailableNotificationServiceImpl.java

```diff
@@ -0,0 +1,369 @@
+        }
+        //todo: find how to get takebyurl link
+        String times = "an unlimited number of";
+        if (!publishedAssessment.getAssessmentAccessControl().getUnlimitedSubmissions()){
+            times = publishedAssessment.getAssessmentAccessControl().getSubmissionsAllowed().toString();
+        }
+        body.append(rl.getFormattedMessage("email.reminder.submit", times));
+        String score = "highest";
+        if (publishedAssessment.getEvaluationModel().getScoringType() == AVERAGE_SCORE){
+            score = "average";
+        } else if (publishedAssessment.getEvaluationModel().getScoringType() == LAST_SCORE){
+            score = "last";
+        } else if(publishedAssessment.getEvaluationModel().getScoringType() == ALL_SCORE){
+            score = "collective";
+        }
+        body.append(rl.getFormattedMessage("email.reminder.score", score));
+        body.append("<br>");
+        String feedback = " ";
+        if(publishedAssessment.getAssessmentFeedback().getFeedbackDelivery() == NO_FEEDBACK){ //3 appears to indicate No Feedback
+            feedback = " no ";
+        }
+        body.append(rl.getFormattedMessage("email.reminder.feedback", feedback));
+        body.append("<br>");
+        body.append(rl.getFormattedMessage("email.reminder.find", site.getTitle(), site.getUrl(), site.getUrl()));
+        Preferences userPrefs = preferencesService.getPreferences(member.getUserId());
+        ResourceProperties samigoOpenNoti = userPrefs.getProperties(NotificationService.PREFS_TYPE + SamigoConstants.NOTI_PREFS_TYPE_SAMIGO_OPEN);    // userPrefs.getProperties("noti:types:sakai:samigoOpen")
+        if(!StringUtils.equals((String) samigoOpenNoti.get("2"), "1")){    //don't send unless the user's preference allows it.
+            emailService.send(fromStr, toStr, subject, body.toString(), headerToStr, replyToStr, additionalHeaders);
+        }
+    }
+
+    private String getReplyTo(Set<Member> instructors) {
+        StringBuilder replyTo = new StringBuilder();
+        for (Member instructor : instructors) {
+            String userEmail = getUserEmail(instructor.getUserId());
+            String displayName = getUserDisplayName(instructor.getUserId());
+            if (StringUtils.isNotEmpty(userEmail)) {
+                replyTo.append("'");
+                replyTo.append(displayName);
+                replyTo.append("' <");
+                replyTo.append(userEmail);
+                replyTo.append(">, ");
+            }
+        }
+        // Return instructors without trailing comma
+        return replyTo.length() > 0 ? replyTo.substring(0, replyTo.length() - 2) : "";
+    }
+
+    private String getUserEmail(String userId) {
+        String email = null;
+        try {
+            email = userDirectoryService.getUser(userId).getEmail();
+        } catch (UserNotDefinedException e) {
+            log.warn("Cannot get email for id: " + userId + " : " + e.getClass() + " : " + e.getMessage());
```

Comment:
Missed this log statement:
```suggestion
            log.warn("Cannot get email for id: {} : {} : {}", userId, e.getClass(), e.getMessage());
```